### PR TITLE
Fix c10 registration binary size

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26865 [wip] Remove manual boxing wrappers
* #26882 Make TypeDefault, TypeDerived and VariableType anonymous namespaces
* #26878 Fix binary size in schema inference
* #26869 Fix shared_ptr binary size in op registration
* #26860 Improve binary size of function schema inference
* **#26827 Fix c10 registration binary size**

The templates there had a binary size impact of ~25MB. This PR fixes that.

Differential Revision: [D17566642](https://our.internmc.facebook.com/intern/diff/D17566642/)

### CODE REVIEW

File: aten/src/ATen/core/op_registration/infer_schema.h [hunk 1]
```diff
@@ -94,7 +94,7 @@ FunctionSchema createFunctionSchemaFromTraits(std::string name, std::string over
 }
 
 template<class FuncType>
-FunctionSchema inferFunctionSchema(std::string name, std::string overload_name) {
+FunctionSchema inferFunctionSchema(std::string&& name, std::string&& overload_name) {
   return detail::createFunctionSchemaFromTraits<guts::infer_function_traits_t<FuncType>>(std::move(name), std::move(overload_name));
```
Comment (resistor) [inline aten/src/ATen/core/op_registration/infer_schema.h]: Are these moves actually doing anything?  Can't we pass this through as a reference all the way to the innermost frame?

Comment (smessmer) [reply aten/src/ATen/core/op_registration/infer_schema.h]: That's why I changed it to `&&`, now it's just passed through as a reference.

Comment (smessmer) [reply aten/src/ATen/core/op_registration/infer_schema.h]: update: changing the `FunctionSchema` constructor reduces by ~100-200kb but that seems within regular build size fluctuation.

Comment (smessmer) [reply aten/src/ATen/core/op_registration/infer_schema.h]: I implemented it in https://github.com/pytorch/pytorch/pull/26860, maybe it helps

**Comment by ezyang** [comment]
Tentatively removing myself as reviewer

**Review by resistor — APPROVED** [review]
Verified the code size benefit

**Pull request closed without merging**